piecrust/src/store/diff.rs

```diff
@@ -10,40 +10,72 @@ use std::io::{Read, Write};
 use qbsdiff::bsdiff::Bsdiff;
 use qbsdiff::bspatch::Bspatch;
 
-use crate::store::mmap::MmapMut;
+use crate::store::memory::Memory;
 
-/// Compute the diff between an `old` and a `new` buffer, and write it to thee
-/// given writer. The length of new buffer is written first, in the form of a
-/// `u64`.
+/// Compute the diff between the old and the new memory, and write it to the
+/// given writer.
+///
+/// First, the difference in size between the old and the new memory will be
+/// written, to handle growth of the memory when [`patch`] is called, and then
+/// the old memory will be grown to match the size of the new memory.
+///
+/// # Important
+/// The old memory should be discarded (dropped) after this operation.
 pub fn diff<T: Write>(
-    old: &[u8],
-    new: &[u8],
+    old: &Memory,
+    new: &Memory,
     writer: &mut T,
-) -> io::Result<u64> {
-    let new_len = new.len() as u64;
-    let new_len_bytes = new_len.to_le_bytes();
+) -> io::Result<()> {
+    let mut old = old.write();
+    let new = new.read();
 
-    writer.write_all(&new_len_bytes)?;
+    let old_len = old.len();
+    let new_len = new.len();
 
-    Bsdiff::new(old, new).compare(writer)
+    debug_assert!(
+        new_len >= old_len,
+        "Length of memories should strictly increase"
+    );
+
+    let delta_len = new_len - old_len;
+    let delta_len_bytes = delta_len.to_le_bytes();
+
+    writer.write_all(&delta_len_bytes)?;
```

Comment:
does it make sense to do this is delta_len is 0 ?

Reply:
We need to know in `patch` if we need to grow the memory or not, so either we put the actual length there, like before, or a diff length.